### PR TITLE
Make symlinks configurable

### PR DESCRIPTION
**Pull request by rahult-graphcore**
This PR makes the symlinking/overlay-creation of datasets more easily extendable (using a config file), and executes the commands from Python rather than in a bash script. 

Previously, we were performing the symlinks in the bash script `prepare-datasets.sh`, and if we had a new dataset to symlink, we'd need to add a new line to the bash script each time. 

After the change, the mapping from `source_dataset_dir` to `target_dir` is defined in a config file `symlink_config.json`. If we need to symlink a new dataset we can just add an entry to the config file. The bash script now calls a python script that reads in the json file and creates the overlays. 

Testing:
1. Create overlays with multiple source dirs to one target dir, and verify in terminal that all files are where they should be (manual on Paperspace)
2. On the main branch vs this feature branch, I wrote out the sorted tree structure of the /tmp directory (the target directory for all the symlinks) to a file, and compared them using `diff` to ensure they were identical. 
3. I ran an automated test to ensure the caches were hit as usual, and these passed: https://graphcore-my.sharepoint.com/personal/alexandrep_graphcore_ai/_layouts/15/onedrive.aspx?ga=1&id=%2Fpersonal%2Falexandrep%5Fgraphcore%5Fai%2FDocuments%2Fpaperspace%2Dtest%2Dreports%2FPhase%2D3%2DSDK3%2E1%2E0%2FPaperspace%2Ftest%2Dreports%2D2023%2D03%2D14

### CODE REVIEW

File: .gradient/prepare-datasets.sh [hunk 1]
```diff
@@ -1,30 +1,8 @@
-#!/bin/bash
-
+#! /usr/bin/env bash 
+set -u 
+set -o pipefail
```
Comment (stuarts-graphcore) [inline .gradient/prepare-datasets.sh:3]: Minor observation, could be written as `set -uxo pipefail`

File: .gradient/prepare-datasets.sh [hunk 2]
```diff
@@ -33,24 +11,17 @@ then
...
-symlink-public-resources "${exe_cache_source_dir}" $POPLAR_EXECUTABLE_CACHE_DIR
-# Symlink squad
-symlink-public-resources "${PUBLIC_DATASET_DIR}/squad" "${HF_DATASETS_CACHE}/squad"
-# Symlink OGB Wiki dataset and checkpoint
-symlink-public-resources "${PUBLIC_DATASET_DIR}/ogbl_wikikg2_custom" "${DATASET_DIR}/ogbl_wikikg2_custom"
-
-# symlink local dataset used by vit-model-training notebook
-symlink-public-resources "${PUBLIC_DATASET_DIR}/chest-xray-nihcc-3" "${DATASET_DIR}/chest-xray-nihcc-3"
+cd "$(dirname "$0")"
+python -u ./symlink_datasets_and_caches.py
```
Comment (stuarts-graphcore) [inline .gradient/prepare-datasets.sh:15]: Do we need to check the return-code of this script?

Also, if an interpreter line is added to the top of the called script, then it can be invoked directly as:

```bash
./symlink_datasets_and_caches.py
```

File: .gradient/prepare-datasets.sh [hunk 3]
```diff
@@ -33,24 +11,17 @@ then
...
+cd -
 
 # pre-install the correct version of optimum for this release
 python -m pip install "optimum-graphcore>=0.5, <0.6"
```
Comment (stuarts-graphcore) [inline .gradient/prepare-datasets.sh:19]: Should this by `python3`?
Suggest checking the return-code also, if we can't operate without this...

File: .gradient/prepare-datasets.sh [hunk 4]
```diff
@@ -33,24 +11,17 @@ then
...
 
 echo "Finished running setup.sh."
 # Run automated test if specified
-if [[ "$1" == "test" ]]; then
+if [[ "${1:-}" == "test" ]]; then
```
Comment (stuarts-graphcore) [inline .gradient/prepare-datasets.sh:23]: Nit-picking to an incredible degree, I like the approach of using single (non-interpolating) quotes for fixed strings, to make it absolutely clear that we're referring to a literal value, e.g.

```bash
if [[ "${1:-}" == 'test' ]]; then
```

File: .gradient/prepare-datasets.sh [hunk 5]
```diff
@@ -33,24 +11,17 @@ then
...
     bash /notebooks/.gradient/automated-test.sh "${@:2}"
```
Comment (stuarts-graphcore) [inline .gradient/prepare-datasets.sh:24]: Unless `automated-test.sh` lacks an initial interpreter line (e.g. `#! /usr/bin/env bash` or similar) or isn't executable then we should invoke it directly, rather than forcing `bash`, e.g. simply:

```bash
/notebooks/.gradient/automated-test.sh "${@:2}"
```

... or, it's possible (although potentially unsafe, if the script uses clashing variables or itself exits) to execute the script in the context of this shell by invoking it as:

```bash
. /notebooks/.gradient/automated-test.sh "${@:2}"
```

(... with `source` being a synonym for `.`) - but this also means that we're using _this_ interpreter rather than the one the invoked script denotes.

File: .gradient/symlink_datasets_and_caches.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import json
```
Comment (stuarts-graphcore) [inline .gradient/symlink_datasets_and_caches.py:1]: Add:

```bash
#! /usr/bin/env python3
```

... to allow the script to be executed directly?

File: .gradient/symlink_config.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+    "$POPLAR_EXECUTABLE_CACHE_DIR":["${PUBLIC_DATASET_DIR}/poplar-executables-pytorch-3-1-1"],
```
Comment (stuarts-graphcore) [inline .gradient/symlink_config.json:2]: ... this line lacks curly-braces around the first key?

**Review by stuarts-graphcore — COMMENTED** [review]
Added some suggestions!

File: .gradient/prepare-datasets.sh (hunk 2 above)
Comment (rahult-graphcore) [reply .gradient/prepare-datasets.sh]: I don't think we need to check the return code. If something fails in the python script, everything else should still continue. It just means that on the Paperspace platform, the poplar executable caches may not be available and so the user would have to compile from scratch. 

Regarding the interpreter line at the top of the python script - from testing, we need to add the `-u` flag to make the output appear without lag on the Paperspace platform, and I couldn't find a way to do that without invoking `python3 -u ...`, so I've kept it that way for now. If there's a way to do that while still invoking directly then I can make that change.

Comment (stuarts-graphcore) [reply .gradient/prepare-datasets.sh]: There's some interesting discussion on this at https://unix.stackexchange.com/questions/399690/multiple-arguments-in-shebang - generally, you can only supply up to two arguments via the interpreter/shebang line... but the coreutils version shipped with Ubuntu 19.04 and above (so, note, not 18.04) would let you do:

```bash
#! /usr/bin/env -S python3 -u
```

... at the start of the script to allow it to be executed directly and still inherit the `-u` option!

Comment (rahult-graphcore) [reply .gradient/prepare-datasets.sh]: It's strange I get a permission error when I try this on Paperspace:
`2023-03-17T10:26:41.818297809Z stderr F /notebooks/.gradient/prepare-datasets.sh: line 13: ./symlink_datasets_and_caches.py: Permission denied`

Comment (rahult-graphcore) [reply .gradient/prepare-datasets.sh]: Updated file permissions, test works now.

File: .gradient/symlink_datasets_and_caches.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import time
+from pathlib import Path
+import subprocess
+import os
+
+
+# read in symlink config file
+with open(f"{Path(__file__).parent.absolute().as_posix()}/symlink_config.json", "r") as f:
+    json_data = f.read()
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:11]: I think we can streamline this line a bit: 
```suggestion
json_data = (Path(__file__).resolve().parent / "symlink_config.json").read_text()
```

File: .gradient/symlink_datasets_and_caches.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+# substitute environment variables in the JSON data
+json_data = os.path.expandvars(json_data)
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:14]: Nice! Didn't know about this

File: .gradient/symlink_datasets_and_caches.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+# parse the json data
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:15]:
```suggestion
```

File: .gradient/symlink_datasets_and_caches.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        # wait until the dataset exists and is populated/non-empty, with a 300s/5m timeout
+        while (COUNTER < 300) and (not source_dir_path.exists() or not any(source_dir_path.iterdir())):
+            print(f"Waiting for dataset {source_dir_path.as_posix()} to be mounted...")
+            time.sleep(1)
+            COUNTER += 1
+
+        # dataset doesn't exist after 300s, so skip it
+        if COUNTER == 300:
+            print(f"Abandoning symlink! - source dataset {source_dir} has not been mounted & populated after 5 minutes.")
+            break
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:35]: Why do you break here? It kind of makes sense, but it means that we might lose out if a dataset which appears later in the list has been mounted: with this break that dataset won't be added to `source_dirs_exist_paths` and won't get symlinked.

(This sort of situation might happen if someone made a typo in one of the dataset names)

File: .gradient/symlink_datasets_and_caches.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        COUNTER = 0
+        # wait until the dataset exists and is populated/non-empty, with a 300s/5m timeout
+        while (COUNTER < 300) and (not source_dir_path.exists() or not any(source_dir_path.iterdir())):
+            print(f"Waiting for dataset {source_dir_path.as_posix()} to be mounted...")
+            time.sleep(1)
+            COUNTER += 1
+
+        # dataset doesn't exist after 300s, so skip it
+        if COUNTER == 300:
+            print(f"Abandoning symlink! - source dataset {source_dir} has not been mounted & populated after 5 minutes.")
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:34]:
```suggestion
            warnings.warn(f"Abandoning symlink! - source dataset {source_dir} has not been mounted & populated after 5 minutes.")
```

Make sure to import warnings

File: .gradient/symlink_datasets_and_caches.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    if len(source_dirs_exist_paths) > 0:
+        print(f"Symlinking - {source_dirs_exist_paths} to {target_dir}")
+        print("-" * 100)
+
+        Path(target_dir).mkdir(parents=True, exist_ok=True)
+
+        workdir_path = Path("/fusedoverlay/workdirs" + target_dir)
+        workdir_path.mkdir(parents=True, exist_ok=True)
+        upperdir_path = Path("/fusedoverlay/upperdir" + target_dir) 
+        upperdir_path.mkdir(parents=True, exist_ok=True)
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:50]: use either dir or path in the variable names, no need for both.

File: .gradient/symlink_datasets_and_caches.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+        lowerdirs = ":".join(source_dirs_exist_paths)
+        overlay_command = f"fuse-overlayfs -o lowerdir={lowerdirs},upperdir={upperdir_path.as_posix()},workdir={workdir_path.as_posix()} {target_dir}"
+        subprocess.run(overlay_command.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:54]: I think you can't have both be PIPE or you might lose some of the output

```suggestion
        subprocess.run(overlay_command.split(), stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
```

File: .gradient/prepare-datasets.sh [hunk 15]
```diff
@@ -33,24 +9,15 @@ then
...
-python -m pip install "optimum-graphcore>=0.5, <0.6"
+python3 -m pip install "optimum-graphcore>=0.5, <0.6"
 
 echo "Finished running setup.sh."
 # Run automated test if specified
-if [[ "$1" == "test" ]]; then
-    bash /notebooks/.gradient/automated-test.sh "${@:2}"
-elif [[ "$2" == "test" ]]; then
-    bash /notebooks/.gradient/automated-test.sh "${@:3}"
+if [[ "${1:-}" == 'test' ]]; then
```
Comment (payoto) [inline .gradient/prepare-datasets.sh:19]: What does this change do?

File: .gradient/symlink_datasets_and_caches.py (hunk 14 above)
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:54]: I feel this could be extracted into a function

File: .gradient/symlink_datasets_and_caches.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+            time.sleep(1)
+            COUNTER += 1
+
+        # dataset doesn't exist after 300s, so skip it
+        if COUNTER == 300:
+            print(f"Abandoning symlink! - source dataset {source_dir} has not been mounted & populated after 5 minutes.")
+            break
+        else:
+            print(f"Found dataset {source_dir}")
+            source_dirs_exist_paths.append(source_dir)
```
Comment (payoto) [inline .gradient/symlink_datasets_and_caches.py:38]: I feel this could be extracted into a function which runs checks for successful mounting?

**Review by payoto — APPROVED** [review]
Looks good, a few small suggestions to make it a bit easier to understand the script

File: .gradient/symlink_datasets_and_caches.py (hunk 11 above)
Comment (rahult-graphcore) [reply .gradient/symlink_datasets_and_caches.py]: Yeah good point, I've removed the break.

**Comment by rahult-graphcore** [comment]
Thanks @payoto, made changes based on your comments. I ran a test on the Pytorch runtime to verify that the caches are symlinked as expected.

**Comment by payoto** [comment]
Sounds good, can you merge?

**Review by payoto — CHANGES_REQUESTED** [review]
Needs to be merged with master, datasets have been added.

**Review by payoto — COMMENTED** [review]
LGTM
